### PR TITLE
[ROCKETMQ-44] Duplicated Codes in DefaultMessageStore.getEarliestMessageTime. https://issues.apache.org/jira/browse/ROCKETMQ-44 || [ROCKETMQ-209] Duplicated code in MQClientApiImpl :- https://issues.apache.org/jira/browse/ROCKETMQ-209

### PR DESCRIPTION
**Pull request by Ritabrata-TW**
This commit is in reference to Story number 44 from the Kanban Board, which refers to removing duplicated logic from DefaultMessageStore.getEarliestMessageTime(). Please let me know if you need any improvements.

### CODE REVIEW

**Comment by Ritabrata-TW** [comment]
@vongosling Please have a look and let me know if you need any further improvements about the same. Thanks.

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 1]
```diff
@@ -745,17 +751,7 @@ public long getMessageStoreTimeStamp(String topic, int queueId, long consumeQueu
...
-                    final long phyOffset = result.getByteBuffer().getLong();
-                    final int size = result.getByteBuffer().getInt();
-                    long storeTime = this.getCommitLog().pickupStoreTimestamp(phyOffset, size);
-                    return storeTime;
-                } catch (Exception ignored) {
-                } finally {
-                    result.release();
-                }
-            }
+            long storeTime = getStoreTime(result);
```
Comment (vsair) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: Convert Long to long ? may lead to NPE?

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 2]
```diff
@@ -717,22 +717,28 @@ public long getEarliestMessageTime(String topic, int queueId) {
...
-                }
-            }
+            Long storeTime = getStoreTime(result);
+            if (storeTime != null) return storeTime;
         }
 
         return -1;
     }
 
+    private Long getStoreTime(SelectMappedBufferResult result) {
```
Comment (vsair) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: IMO, just return long and default value is -1.

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java [hunk 3]
```diff
@@ -1215,9 +1199,14 @@ public TopicRouteData getTopicRouteInfoFromNameServer(final String topic, final
         assert response != null;
         switch (response.getCode()) {
             case ResponseCode.TOPIC_NOT_EXIST: {
-                if (!topic.equals(MixAll.DEFAULT_TOPIC))
-                    log.warn("get Topic [{}] RouteInfoFromNameServer is not exist value", topic);
-                break;
+                if(allowTopicNotExist) {
```
Comment (vsair) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: Could you polish the 'double if'?

**Comment by Ritabrata-TW** [comment]
Hi @vsair . Thanks for the suggestions. I have incorporated the same. Please have a look. Thanks

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 4]
```diff
@@ -16,29 +16,7 @@
...
-import java.util.concurrent.ScheduledExecutorService;
-import java.util.concurrent.TimeUnit;
-import java.util.concurrent.atomic.AtomicLong;
-import org.apache.rocketmq.common.BrokerConfig;
-import org.apache.rocketmq.common.MixAll;
-import org.apache.rocketmq.common.ServiceThread;
-import org.apache.rocketmq.common.SystemClock;
-import org.apache.rocketmq.common.ThreadFactoryImpl;
-import org.apache.rocketmq.common.UtilAll;
+import org.apache.rocketmq.common.*;
```
Comment (shroman) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: I believe you won't have this * if you format promerly.
Please check http://rocketmq.apache.org/docs/code-guidelines/

Comment (Ritabrata-TW) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: @shroman Done. Thanks for pointing towards the code guidelines.

**Comment by Ritabrata-TW** [comment]
@shroman Can you take a look? Thanks

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java [hunk 5]
```diff
@@ -1215,9 +1210,13 @@ public TopicRouteData getTopicRouteInfoFromNameServer(final String topic, final
         assert response != null;
         switch (response.getCode()) {
             case ResponseCode.TOPIC_NOT_EXIST: {
-                if (!topic.equals(MixAll.DEFAULT_TOPIC))
+                if (allowTopicNotExist && !topic.equals(MixAll.DEFAULT_TOPIC)) {
                     log.warn("get Topic [{}] RouteInfoFromNameServer is not exist value", topic);
-                break;
+                    break;
```
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: why do you need 2 breaks?

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java [hunk 6]
```diff
@@ -1215,9 +1210,13 @@ public TopicRouteData getTopicRouteInfoFromNameServer(final String topic, final
...
+                } else {
+                    // TODO LOG
```
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: Please don't leave TODO if possible.

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java [hunk 7]
```diff
@@ -717,22 +718,28 @@ public long getEarliestMessageTime(String topic, int queueId) {
...
-                    final int size = result.getByteBuffer().getInt();
-                    long storeTime = this.getCommitLog().pickupStoreTimestamp(phyOffset, size);
-                    return storeTime;
-                } catch (Exception e) {
-                } finally {
-                    result.release();
-                }
-            }
+            Long storeTime = getStoreTime(result);
+            return storeTime;
```
Comment (shroman) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: `return getStoreTime(result);` will suffice.

**Comment by vongosling** [comment]
@Ritabrata-TW Could you modify PR subject conforming to our norm :-)

**Comment by vongosling** [comment]
@Ritabrata-TW I am not understand your refer "Story number 44 from the Kanban Board", could you detail it ?

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java (hunk 1 above)
Comment (vongosling) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: We appreciate your contribution for rocketmq commuity, could you use unit test to verify your refactoring

**Comment by shroman** [comment]
@Ritabrata-TW In other words, your pull request description has to be in this format
```
[ROCKETMQ-xxx] Message describing the issue.

Link to ROCKETMQ-xxx, which is the JIRA issue.
```

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 5 above)
Comment (Ritabrata-TW) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: @shroman We don't have two breaks. The part you are have referred to shows the diff where the break from 1220 has been removed and the break on 1215 has been added.

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 6 above)
Comment (Ritabrata-TW) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: @shroman The TODO was already there. Not sure what to do about it right now.

File: store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java (hunk 7 above)
Comment (Ritabrata-TW) [inline store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java]: Yup. Made the change.

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 6 above)
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: ok then

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 5 above)
Comment (shroman) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: well, I mean you can have one for `case ResponseCode.TOPIC_NOT_EXIST`, don't you?

**Comment by vongosling** [comment]
@Ritabrata-TW Looking forward to your following for this PR :-)

File: client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java (hunk 5 above)
Comment (Ritabrata-TW) [inline client/src/main/java/org/apache/rocketmq/client/impl/MQClientAPIImpl.java]: @shroman Ahh. I get you now. Making the change and pushing. Thanks

**Comment by Ritabrata-TW** [comment]
@vongosling Sorry for the late response. I have changed the PR description according to the format suggested. I agree to what you said about splitting the big refactor into multiple parts and will keep this in mind for future issues. This PR has commits from 2 issues and it would be a little tedious to pull out these commits into multiple PRs now. I will attach both the issues in the PR description. Please have a look and let me know if this is okay.

**Comment by vongosling** [comment]
@Ritabrata-TW thanks, I will take a close look it later :-)

**Comment by vsair** [comment]
Is there something wrong with your code style ?

**Comment by zhouxinyu** [comment]
@Ritabrata-TW Hi, thanks for your contribution, and could you please set your merge target to `develop` branch which is our daily development branch?

**Comment by Ritabrata-TW** [comment]
@vsair The Travis build seems to have passed. Did you notice something wrong with the code style ?

**Comment by Ritabrata-TW** [comment]
@zhouxinyu My pleasure. I have set the merge target to develop. Thanks.

**Comment by Ritabrata-TW** [comment]
@vongosling @zhouxinyu Anything else that I need to do on this? Otherwise, can you please have a look and merge it?

**Comment by vongosling** [comment]
@Ritabrata-TW I found so many code format in your PR, could you import our code style file as instruction, http://rocketmq.incubator.apache.org/docs/code-guidelines/

**Comment by zhouxinyu** [comment]
Thanks @Ritabrata-TW , we will merge this PR soon, but please don't forget to add related unit tests and follow our [code style](http://rocketmq.incubator.apache.org/docs/code-guidelines/) next time.
You are welcome to continue keeping your attention on RocketMQ community.

**Comment by Ritabrata-TW** [comment]
@zhouxinyu Thanks. I will keep the suggested things in mind next time. Please let me know if you need anything to close the PR.

I would also like to get more involved with the RocketMQ community and would be dropping a mentoring request in the mailing list. Thanks.

**Comment by zhouxinyu** [comment]
@Ritabrata-TW this PR has been merged into `develop` branch and will be released in next version, you can close it now safely.

**Comment by Ritabrata-TW** [comment]
@zhouxinyu Thanks a lot. Will look forward to working with you again.

**Comment by vongosling** [comment]
@Ritabrata-TW Thanks for your attention for rocketmq community, whats' your scenario when using apache rocketmq in your company :-)

**Comment by Ritabrata-TW** [comment]
@vongosling We do not need a messaging queue in our Project as of yet. Will be advocating for RocketMQ if the scenario arises.

**Pull request closed without merging**